File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    largestRectangleArea(heights) {
        let maxArea = 0;
        let stack = [];

        for (let i = 0; i < heights.length; i++) {
            const h = heights[i];
            let start = i;

            while (stack.length > 0 && stack[stack.length - 1][1] > h) {
                const [popIdx, popH] = stack.pop();
                maxArea = Math.max(maxArea, popH * (i - popIdx));
                start = popIdx;
            }
            stack.push([start, h]);
        }

        while (stack.length > 0) {
            const [popIdx, popH] = stack.pop();
            const width = heights.length - popIdx;
            maxArea = Math.max(maxArea, popH * width);
        }

        return maxArea;
    }
}

File: Data Structures & Algorithms/longest-turbulent-subarray/submission-0.java
class Solution {
    public int maxTurbulenceSize(int[] arr) {
        int l = 0;
        int r = 1;
        int res = 1;
        char prev = '=';

        while (r < arr.length) {
            if (arr[r - 1] > arr[r] && prev != '>') {
                res = Math.max(res, r - l + 1);
                r += 1;
                prev = '>';
            } else if (arr[r - 1] < arr[r] && prev != '<') {
                res = Math.max(res, r - l + 1);
                r += 1;
                prev = '<';
            } else {
                if (arr[r] == arr[r - 1]) {
                    r = r + 1;
                }
                l = r - 1;
                prev = '=';
            }
        }
        return res;
    }
}